src/uucore/src/lib/features/checksum.rs

```diff
@@ -2,26 +2,967 @@
 //
 // For the full copyright and license information, please view the LICENSE
 // file that was distributed with this source code.
+// spell-checker:ignore anotherfile invalidchecksum
 
-use crate::show_warning_caps;
+use os_display::Quotable;
+use regex::Regex;
+use std::{
+    error::Error,
+    ffi::OsStr,
+    fmt::Display,
+    fs::File,
+    io::{self, BufReader, Read},
+    path::Path,
+};
+
+use crate::{
+    error::{set_exit_code, FromIo, UError, UResult, USimpleError},
+    show, show_error, show_warning_caps,
+    sum::{
+        Blake2b, Blake3, Digest, DigestWriter, Md5, Sha1, Sha224, Sha256, Sha384, Sha3_224,
+        Sha3_256, Sha3_384, Sha3_512, Sha512, Shake128, Shake256, Sm3, BSD, CRC, SYSV,
+    },
+    util_name,
+};
+use std::io::stdin;
+use std::io::BufRead;
+
+pub const ALGORITHM_OPTIONS_SYSV: &str = "sysv";
+pub const ALGORITHM_OPTIONS_BSD: &str = "bsd";
+pub const ALGORITHM_OPTIONS_CRC: &str = "crc";
+pub const ALGORITHM_OPTIONS_MD5: &str = "md5";
+pub const ALGORITHM_OPTIONS_SHA1: &str = "sha1";
+pub const ALGORITHM_OPTIONS_SHA3: &str = "sha3";
+
+pub const ALGORITHM_OPTIONS_SHA224: &str = "sha224";
+pub const ALGORITHM_OPTIONS_SHA256: &str = "sha256";
+pub const ALGORITHM_OPTIONS_SHA384: &str = "sha384";
+pub const ALGORITHM_OPTIONS_SHA512: &str = "sha512";
+pub const ALGORITHM_OPTIONS_BLAKE2B: &str = "blake2b";
+pub const ALGORITHM_OPTIONS_BLAKE3: &str = "blake3";
+pub const ALGORITHM_OPTIONS_SM3: &str = "sm3";
+pub const ALGORITHM_OPTIONS_SHAKE128: &str = "shake128";
+pub const ALGORITHM_OPTIONS_SHAKE256: &str = "shake256";
+
+pub const SUPPORTED_ALGO: [&str; 15] = [
+    ALGORITHM_OPTIONS_SYSV,
+    ALGORITHM_OPTIONS_BSD,
+    ALGORITHM_OPTIONS_CRC,
+    ALGORITHM_OPTIONS_MD5,
+    ALGORITHM_OPTIONS_SHA1,
+    ALGORITHM_OPTIONS_SHA3,
+    ALGORITHM_OPTIONS_SHA224,
+    ALGORITHM_OPTIONS_SHA256,
+    ALGORITHM_OPTIONS_SHA384,
+    ALGORITHM_OPTIONS_SHA512,
+    ALGORITHM_OPTIONS_BLAKE2B,
+    ALGORITHM_OPTIONS_BLAKE3,
+    ALGORITHM_OPTIONS_SM3,
+    ALGORITHM_OPTIONS_SHAKE128,
+    ALGORITHM_OPTIONS_SHAKE256,
+];
+
+pub struct HashAlgorithm {
+    pub name: &'static str,
+    pub create_fn: Box<dyn Fn() -> Box<dyn Digest + 'static>>,
+    pub bits: usize,
+}
+
+#[derive(Debug)]
+pub enum ChecksumError {
+    RawMultipleFiles,
+    IgnoreNotCheck,
+    InvalidOutputSizeForSha3,
+    BitsRequiredForSha3,
+    BitsRequiredForShake128,
+    BitsRequiredForShake256,
+    UnknownAlgorithm,
+    InvalidLength,
+    LengthOnlyForBlake2b,
+    BinaryTextConflict,
+    AlgorithmNotSupportedWithCheck,
+    CombineMultipleAlgorithms,
+    NeedAlgoToHash,
```

Comment:
For consistency reasons I would use `NeedAlgorithmToHash` because the other variants of this enum use "Algorithm", and not "Algo".